src/typeInference/annotateAST.ts

```diff
@@ -21,10 +24,14 @@ export default function annotateAst(ast: IAST, staticContext?: StaticContext) {
 }
 
 /**
- * This is the recursive function used to annotate any AST node.
- * @param ast The AST to annotate
- * @param staticContext The static constext to use for function lookups
- * @returns The type of the AST node
+ * Recursively traverse the AST in the depth first, pre-order to infer type and annotate AST;
+ * Annotates as much type information as possible to the AST nodes.
+ * Inserts attribute `type` to the corresponding node if type is inferred.
+ *
+ * @param ast The AST to annotate.
+ * @param staticContext The static context to use for function lookups.
+ * @throws errors when attempts to annotate fail.
```

Comment:
Right now this function never throws an error. If we cannot annotate a node we return undefined.